msprime/simulations.py

```diff
@@ -1365,6 +1368,139 @@ def print_history(self, output=sys.stdout):
+
+    def coalescence_rate_trajectory(self, end, num_samples,
+                                    num_steps=10, min_pop_size=1):
+        """
+        This function will calculate the ground truth
+        coalescent rate trajectory, r (Ne = 1 / 2r), for a population with
+        multiple sub-populations. This takes a
+        demographic model and returns a
+        function of time that gives haploid coalescence rate.
+
+        `num_samples` should be a list of the same length as the number
+        of populations, so that num_samples[j] is the number of
+        samples in subpopulation j.
+
+        `num_steps` parameter will determine how many points along the
+        time axis are returned.
+
+        `min_pop_size` is the smallest a population size can be represented as during
+        the computation of coalescent rates. In models where populations
+        grow exponentially, very small effective population sizes skew the
+        calculation when we want to compute coalescent rates within a population (1/2Ne).
+
+        Returns a tuple where the first value
+        is the steps (in generations) and the second
+        is an array whose jth element is the coalescence rate at time
+        steps[j], i.e., the probability per unit time that a randomly
+        chosen pair of lineages from the original samples coalesces
+        around that time, conditioned on the pair not having yet coalesced.
+        The third is the probablility that the pair of lineages
+        has not yet coalesced.
+        """
+
+        num_pops = self.num_populations
+        assert(len(num_samples) == num_pops)
+        P = np.zeros([num_pops**2, num_pops**2])
+        IA = np.array(range(num_pops**2)).reshape([num_pops, num_pops])
+        Identity = np.eye(num_pops)
+        for x in range(num_pops):
+            for y in range(num_pops):
+                K_x = num_samples[x]
+                K_y = num_samples[y]
+                P[IA[x, y], IA[x, y]] = K_x * (K_y - (x == y))
+        P = P / np.sum(P)
+        r = np.zeros(num_steps)
+        p_t = np.zeros(num_steps)
+        steps = np.linspace(0, end, num_steps)
+        dt = steps[1] - steps[0]
+        mass_migration_objects = []
+        mass_migration_steps = []
+        for demo in self.demographic_events:
+            if type(demo) == MassMigration:
+                mass_migration_objects.append(demo)
+                mass_migration_steps.append(int(demo.time/dt))
+        for j in range(num_steps):
+            N_M = self.pop_size_and_migration_at_t(steps[j])
+            N = N_M["population_sizes"]
+            M = copy.deepcopy(N_M["migration_matrix"])
+            C = np.zeros([num_pops**2, num_pops**2])
+            for idx in range(num_pops):
+                C[IA[idx, idx], IA[idx, idx]] = 1 / (2 * max(min_pop_size, N[idx]))
+            for idx, row in enumerate(M):
+                M[idx][idx] = -1 * sum(row)
+            G = (np.kron(M, Identity) + np.kron(Identity, M)) - C
```

Comment:
we could skip the deepcopy (trying to avoid extra dependencies here) by doing the following thing instead of lines 1457-1459:
```
             dM = np.diag(sum(M))
             G = (np.kron(M - dM, Identity) + np.kron(Identity, M - dM)) - C
```

Reply:
oh, right, much easier